### ISSUE TITLE
Installfest Vagrant VM proposal

### ISSUE THREAD

**Issue by danchoi** [railsbridge-boston/railsbridge-boston#86]
I started using Vagrant for our own business. And it dawned on me that Vagrant is a much faster and more reliable way install all the pre-requisites for the Railsbridge workshop. It is simpler and more uniform than RailsInstaller and Tokaido. This is how the Vagrant-based install process would work:

We pre-bake a Vagrant box, which is a VM disk image, with an Ubuntu 12 server and all the dependencies for the Railsbridge tutorial pre-installed. 

Then at the workshop, students will only need to the do the following:
1. Download and install VirtualBox using the VirtualBox installer (OS X, Windows, and Linux available)
2. Download and install Vagrant using the Vagrant installer (same 3 platforms covered).
3. Download our Railsbridge Boston Vagrant box & start the virtual machine.

And that is basically it. Students can get to the command line in the VM with `vagrant ssh` & immediately start doing Ruby and git pushes and rake invocations. The gems will already be installed.

As for the text editor, Vagrant sets up a directory that is shared between the host system and the VM, so students will be able to edit their source code just like they have in the past, using whatever editor they want. For command line tasks, everyone will be in the Ubuntu environment, and we will be spared weird Windows command line issues. We won't have to deal with XCode anymore either.

I know this is a fairly aggressive change I'm proposing, but I can set up a Railsbridge Vagrant VM this week so anyone can try out this this new install and setup process  to verify that it works well. I think the total Installfest setup time will be reduced to less than 20 minutes & there will be fewer problematic outliers.

**Comment by danchoi** [comment]
I just remembered that @dpickett recommended using Vagrant for the installfest back in May of 2012 during the planning for the 1st BostonRB Project Night -- especially to avoid Windows installation headaches.

**Comment by dpickett** [comment]
yeah I think this is a neat idea. For me, it comes down to the objectives of the workshop - do we want to show them how much they can do with ruby and rails or do we want to get them installed and configured so they can explore on their own? There's something to be said for the latter, but my feeling is that an emphasis on what can be done with the framework would serve attendees better.

I think we could provide the instructions for local installation either before or after and encourage them to come to a BostonRB project night or other venue to complete that should they get stuck, but wouldn't it be awesome if we could spend the whole day and a half experimenting and building instead of configuring?

**Comment by slothbear** [comment]
I'll try it out. :+1:

**Comment by danchoi** [comment]
OK v1 is ready to test:

https://github.com/railsbridge-boston/railsbridge-virtual-machine

**Comment by drapergeek** [comment]
I think this would be a neat idea but I also feel like it is introducing more complexity than has been said so far. Vagrant introduces another layer of confusion for where you enter commands and where you open directories to work in. I love vagrant but I personally think that this would introduce another layer of confusion for people in where they work.

**Comment by danchoi** [comment]
Would you agree at least that a custom Vagrant VM simplifies things for
Windows users? The Git Bash option, prompt issues, and a bunch other issues
go away.

Also, RailsInstaller doesn't support Ruby 2.0 or Rails 4.0 yet. So we can
wait or try to patch RailsInstaller or go back to a more manual
installation method for OS X. Or we can use the custom Vagrant VM.

**Comment by drapergeek** [comment]
I don't consider windows users when thinking about anything :smile: 

Jokes aside, I really don't know how easy it is to use vagrant on windows or what is used to connect so I can't provide an opinion there, only on OS X. It may be the simpler decision on Windows.

I don't personally think it is worth it to force the use of Rails 4.0 and ruby 2.0 for this next workshop, just wait until we're on the same version with RailsInstaller.

**Comment by danchoi** [comment]
OK I'll leave it to you guys to decide.

**Comment by drapergeek** [comment]
I wouldn't be opposed to having a vagrant setup for windows and a non vagrant for others, that would be an excellent A/B test!

**Comment by danchoi** [comment]
Sounds good to me!

**Comment by danchoi** [comment]
I'd like to test the Vagrant install on pre OS X 10.7 users as well.

**Comment by drapergeek** [comment]
That also seems reasonable.

**Comment by danchoi** [comment]
Sounds good.

FYI about future RailsInstaller support for Rails 4

https://github.com/railsinstaller/railsinstaller-nix/issues/20

**Comment by bcarreno** [comment]
I like the idea.

In the past having an environment the students can use going forward has been an argument against this approach.  But as Dan Pickett says, we can address that at project nights.

I'll test it as soon as I have the chance.

**Comment by slothbear** [comment]
I don't understand how the Vagrant approach leads to more confusion. Students are already have the confusion of multiple terminal and editor windows. Perhaps the `vagrant ssh` command line could have some beautiful colorful RBB flourish to distinguish the place to enter commands. Whether or not this approach adds confusion, it solves many troubles that students (and TAs) run into. I had to ask 3 different TAs last time before someone knew about `rbenv rehash`. I'm pretty sure the student didn't learn anything useful in that interaction.

I also don't understand (wow, I'm big on not understanding tonight) how the Vagrant approach precludes students using the environment going forward. With some documentation, students will figure out how to stop and start the Vagrant system.

I'm interested in how the setup works on older systems. Do we have any contacts in Montreal who could report on how things went for them when they tried this approach? Their instructions included a Linux GUI (which sounds crazy) but it would be interesting to hear how well the students did with this approach.

**Comment by tdyer** [comment]
I can see that using Vagrant on OSX is _not_ the typical development setup. I've mostly used Vagrant to mock a Linux production environment. And I think that students being taught using Vagrant on this platform would want, at some point, to use the typical OSX setup. Perhaps, it would discourage them at some point down the road? 

I'm a newbie to railsbridge, only TA'd at the last session, so I don't really know the kinds of pains you guys have seen with config and install. 

Thankfully, I haven't done any Ruby/Rails development on Windows for years. It was really painful back then and I'm not familiar enough with the railsinstaller to judge if Vagrant would be a better option on Windows. I think I could get my hands on a window box or two to help out with the Vagrant install process if that's the decision.

**Comment by danchoi** [comment]
For me, the dream is to eliminate the Friday night session by making the
installation as fast and reliable as possible. Based on a few initial
tests, the custom Vagrant VM may be able to reduce the installfest to less
than 20 minutes with no or very few problem cases.

It's true that RailsInstaller has sped up the installfest for most OS X and
Windows users. However, RailsInstaller may be frozen at Rails 3.2 due to
legal issues surrounding bundling the Apple gcc compiler or whatnot. That
would mean that going forward, we would have to revert to having people
install XCode and then run a RVM install script which in turn bundle
installs all the necessary gems. This is more time and bandwidth consuming
and glitch-prone -- enough to make the Friday night session absolutely
unavoidable.

An extra night adds about 25% to the cost of running the workshop along
with higher difficulty in securing venue and a greater scheduling burden on
volunteers and attendees -- especially those with young children who need
to schedule babysitters. It would be nice if we could save that 25% and buy
t-shirts with it, and also alleviate the scheduling challenge for a lot of
people.

I appreciate that working inside a Vagrant VM isn't as "natural" as working
in your own native environment. But the Railsbridge workshop is if anything
about striking smart compromises. We might lose a little on one dimension
but gain something very worthwhile on some others, like making attending
easier for a non-trivial number of people. I think that's the way I would
frame the issue here.

**Comment by slothbear** [comment]
I don't understand ( _3rd time tonight_ ) how working with a Vagrant VM is any different than the native environment. There are a couple of startup and shutdown commands. After that, vagrant ssh is a command line – probably in the same terminal you'd use in your native environment. You use whatever native editor & browser you like. What's unnatural? I got the sample VM up and running. Perhaps I need to take it through the whole curriculum to understand the snags.

I hadn't thought about the possibility of skipping the Installfest altogether. Interesting idea.

**Comment by tdyer** [comment]
Sounds like Vagrant isn't different enough, more typical OSX installs could happen at a later project night and Dan makes great points about the value of reducing/eliminating the installfest. 

Could a "virtual" installfest help for those having problems? Perhaps one or two nights in a google hangout, chat, ...? Maybe virtual installfest/s that are targeted at only one platform?

**Comment by danchoi** [comment]
I think @drapergeek is right to be conservative for the next installfest, but for the longer term I would like to propose that we try to have everyone self-install the stuff before the workshop. They could come to a small, optional pre-workshop installation help session staffed by a small number of volunteers if they fail someone along the line. We could also offer chat, do a instructional screencast video, etc. I think a one-day workshop is worth trying to get to.

**Comment by bcarreno** [comment]
I've tested the installation on Mac and it looks good.

@kgf has offered to try it on Linux

**Comment by bcarreno** [comment]
Thank you @kgf for your help!

**Comment by bcarreno** [comment]
@tdyer good ideas, once we try vagrant in the next workshop we can see what we need.

@tdyer are you still interested in being in charge of the installfest for the next workshop? in this case, do you think you can test the procedure on Windows and take notes? (we need to update the instructions).  I don't think anyone has tested Windows yet.

**Comment by danchoi** [comment]
@bcarreno @drapergeek Are we agreed on the following:
1. RailsInstaller for supported OS X versions, per Jason Draper's comments
2. Vagrant for non-supported OS X and Windows, in order to experiment and test the potential of Vagrant installs
3. Linux -- undecided which way to go

Also, non-Intel laptops cannot run VirtualBox -- so that rules out AMD, PowerPC, and probably Atom. I suggest we try to know beforehand which attendees are in this situation with questions in the application & then assign TAs to be prepared to help these people.

**Comment by bcarreno** [comment]
@slothbear Beginners can get confused when they're in a different environment or "mode".   We've seen this when they're inside irb and they don't understand why this is different than the command line.  They'll be in a ssh session connected to a virtual machine that's living in the same laptop, so I agree this can create confusion.  However, the benefits greatly outweigh this problem, if we can eliminate Friday it would be a huge improvement.

**Comment by danchoi** [comment]
We may also have to create a new Vagrant box to targets Rails 3.2, since that is the highest Rails that RailsInstaller supports. The curriculum should also perhaps target Rails 3.2.

**Comment by tdyer** [comment]
@bcarreno Yes, I do want to do the installfest for the next workshop. I'll get a hold of a Windows box and do the Vagrant install. Probably, will do over this coming weekend and update the instructions.

**Comment by bcarreno** [comment]
@danchoi agreed.  I think it's better to have everyone on the same version, i.e. Rails 3.2

@tdyer thank you so much!

**Comment by slothbear** [comment]
I gave the installation a try on Windows 7 Home Premium. It took me a little less than an hour. Downloading Virtual Box, Vagrant (with restart) and the rbb box took about 20 minutes.The other big time sink was getting something to replace `vagrant ssh` since my Windows had no SSH client.

I took advice from Lifehacker to use PuTTY, then I had to figure out how to connect it to the virtual machine. For the record:
- When you run `vagrant ssh` on Windows, instructions are given for connecting via a different client (host and port).
- After opening a session with the given host and port, login with _username_: **vagrant**, _password_: **vagrant**. (many sources talked about using PuTTYGen to convert keys, but I didn't need to do that. The password worked fine.)
- The `vagrant ssh` instructions mentioned Cygwin, MinGW, or Git as alternative SSH clients. Git sounds interesting and I got it to work -- but installing Windows Git just to get an SSH client seems like a lot of work.
